fuzz/fuzz_types.ml

```diff
@@ -0,0 +1,440 @@
+ fun record_name field_name field_type dynrec ->
+  let curr_record_name, curr_table = dynrec in
+  if not (String.equal record_name curr_record_name) then
+    failwith "Trying to access fields from the wrong record."
+  else
+    let wrapped = Hashtbl.find curr_table field_name in
+    unwrap field_type wrapped
+
+(** Dynamically build an Irmin variant (assuming it has 1 to 5 cases). *)
+and irmin_variant : string -> case list -> dyn_variant T.t = [%impl_variant 5]
+
+(** Dynamically build an Irmin enum (assuming it has 1 to 5 cases). *)
+and irmin_enum : string -> string list -> dyn_variant T.t =
+ fun n ts -> irmin_variant n (List.map (fun t -> (t, ACT Case0)) ts)
+
+(** Create a generator for values of a given dynamic type. *)
+let rec t_to_value_gen : type a. a t -> a gen = function
+  | TUnit -> const ()
+  | TBool -> bool
+  | TChar -> char
+  | TInt -> int
+  | TInt32 -> int32
+  | TInt64 -> int64
+  | TFloat -> float
+  | TString -> string
+  | TBytes -> bytes
+  | TList ty -> list (t_to_value_gen ty)
+  | TArray ty -> map [ list (t_to_value_gen ty) ] Array.of_list
+  | TOption ty -> option (t_to_value_gen ty)
+  | TPair (ty1, ty2) -> pair (t_to_value_gen ty1) (t_to_value_gen ty2)
+  | TTriple (ty1, ty2, ty3) ->
+      triple (t_to_value_gen ty1) (t_to_value_gen ty2) (t_to_value_gen ty3)
+  | TResult (ty1, ty2) -> result (t_to_value_gen ty1) (t_to_value_gen ty2)
+  | TRecord (n, fs) -> irmin_record_gen n fs
+  | TVariant (n, cs) -> irmin_variant_gen n cs
+  | TEnum (n, ts) -> irmin_enum_gen n ts
+
+(** Create a generator for an Irmin record with the given specification. *)
+and irmin_record_gen : string -> field list -> dyn_record gen =
+ fun record_name fields ->
+  let rec fields_gen = function
+    | [] -> const []
+    | (field_name, AT field_type) :: xs ->
+        map [ t_to_value_gen field_type; fields_gen xs ] (fun v fs' ->
+            (field_name, wrap field_type v) :: fs')
+  in
+  map [ fields_gen fields ] (fun fs' -> new_dyn_record record_name fs')
+
+(** Create a generator for an Irmin variant with the given specification. *)
+and irmin_variant_gen : string -> case list -> dyn_variant gen =
+ fun variant_name cases ->
+  let rec cases_gen = function
+    | [] -> []
+    | (case_name, ACT Case0) :: xs ->
+        const (variant_name, case_name, VUnit ()) :: cases_gen xs
+    | (case_name, ACT (Case1 t)) :: xs ->
+        map [ t_to_value_gen t ] (fun v -> (variant_name, case_name, wrap t v))
+        :: cases_gen xs
+  in
+  choose (cases_gen cases)
+
+(** Create a generator for an Irmin enum with the given specification. *)
+and irmin_enum_gen : string -> string list -> dyn_variant gen =
+ fun n ts -> irmin_variant_gen n (List.map (fun t -> (t, ACT Case0)) ts)
+
+type inhabited = Inhabited : 'a t * 'a -> inhabited
+
+(** Generate a [Inhabited (t, v)], where [t : 'a t] is a dynamic type and
+    [v : 'a] is a corresponding value. *)
+let inhabited_gen : inhabited gen =
+  dynamic_bind any_type_gen @@ fun (AT t) ->
```

Comment:
There is a warning about [dynamic_bind](https://github.com/stedolan/crowbar/blob/master/src/crowbar.mli#L166). Do you know what "optimizations or fuzzing techniques" we might be missing out? An alternative is to have a test for each type of irmin, instead of generating the types as well, but it would not be so elegant.

Reply:
I don't think we really have a choice here. For us to be able to test arbitrary types, we must first generate a type and then a value of that type; and since there are infinitely many Irmin types, I don't see how we could have "a test for each type".